[WFLY-17021]: Upgrade jboss metadata to 15.2.0.Beta3

ehsavoie opened the pull request:
* Adding jboss-ejb3-spec-4_0.xsd
* Adding jboss-ejb-iiop_2_0.xsd
* Adding trans-timeout-2_0.xsd

Jira: https://issues.redhat.com/browse/WFLY-17021

Signed-off-by: Emmanuel Hugonnet <ehugonne@redhat.com>

ehsavoie committed:
[WFLY-17021]: Upgrade jboss metadata to 15.2.0.Beta3

* Fixing some XML descriptors.
* Updating jboss-metadata to 15.2.0.Beta3:
	- Adding jboss-ejb3-spec-4_0.xsd
	- Adding jboss-ejb-iiop_2_0.xsd
	- Adding trans-timeout-2_0.xsd

Jira: https://issues.redhat.com/browse/WFLY-17021

Signed-off-by: Emmanuel Hugonnet <ehugonne@redhat.com>

bstansberry commented:
@ehsavoie This shows a bunch of failures.

chengfang commented:
looks like all the failures are related to ejb security. I suspect we also need to update the various namespaces in this schema:
https://github.com/wildfly/wildfly/blob/main/ejb3/src/main/resources/schema/jboss-ejb-security_1_1.xsd

ehsavoie commented:
@chengfang ok, i'll update all those schemas in a second commit.

ehsavoie committed:
[WFLY-17025,WFLY-17026,WFLY-17027,WFLY-17028,WFLY-17029,WFLY-17030,WFLY-17031,WFLY-17032]: Upgrade EJB3 schema that are nto JakartaEE 10 compatible.

Make the following schemas jakartaEE10 compatible:
 - jboss-ejb-security
 - jboss-ejb-security-role
 - jboss-ejb-timer-service
 - jboss-ejb-resource-adapter-binding
 - jboss-ejb-pool
 - jboss-ejb-delivery-active
 - jboss-ejb-container-interceptors
 - jboss-ejb-clustering
 - jboss-ejb-cache

Jira:
* https://issues.redhat.com/browse/WFLY-17025
* https://issues.redhat.com/browse/WFLY-17026
* https://issues.redhat.com/browse/WFLY-17027
* https://issues.redhat.com/browse/WFLY-17028
* https://issues.redhat.com/browse/WFLY-17029
* https://issues.redhat.com/browse/WFLY-17030
* https://issues.redhat.com/browse/WFLY-17031
* https://issues.redhat.com/browse/WFLY-17032

Signed-off-by: Emmanuel Hugonnet <ehugonne@redhat.com>

ehsavoie committed:
[WFLY-17021)]: Updating some descriptors to JakartaEE 10 namespace

Signed-off-by: Emmanuel Hugonnet <ehugonne@redhat.com>

chengfang commented on ejb3/src/main/java/org/jboss/as/ejb3/deliveryactive/parser/EJBBoundMdbDeliveryMetaDataParser20.java:
```diff
@@ -0,0 +1,87 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
New files have Apache license, which I think should be LGPL in WildFly.

chengfang commented on testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/messaging/jms/definitions/JMSResourceManagementTestCase.java:
```diff
@@ -75,9 +77,12 @@ public static JavaArchive createArchive() {
+"</beans>"),
                         create("beans.xml"));
-
+        archive.as(ZipExporter.class).exportTo(new File("JMSResourceDefinitionsTestCase.jar"));
```
is this a temporary debugging statement?

ehsavoie replied:
yes

ehsavoie replied:
correct

chengfang commented on ejb3/src/main/resources/schema/jboss-ejb-timer-service_3_0.xsd:
```diff
@@ -0,0 +1,56 @@
+                            <xs:documentation>Identifies the timer-management provider to use for persistent timers.</xs:documentation>
+                        </xs:annotation>
+                    </xs:element>
+                </xs:choice>
```
Compared to previous version (https://github.com/wildfly/wildfly/blob/main/ejb3/src/main/resources/schema/jboss-ejb-timer-service_2_0.xsd), the structure of this element is different.

In existing version, there is a:
all
    choice (config for persistent timer)
        option 1 for persistent timer
        option 2 for persistent timer
    choice (config for non-persistent timer)

In the new version, it's flattened to a choice element. Any reason for this change?

chengfang replied:
Also looks like the parser for this xml is not updated (TimerServiceMetaDataParser and TimerServiceMetaDataSchema)

ehsavoie replied:
the xs:all can't have a xs:choice as child

chengfang replied:
@pferraro fyi

pferraro replied:
Yeah, that schema is messed up.  It should look something like:
```xml
    <xs:complexType name="timer-serviceType">
        <xs:complexContent>
            <xs:extension base="javaee:jboss-assembly-descriptor-bean-entryType">
                <xs:choice>
                    <xs:element name="persistence-store-name" type="xs:string">
                        <xs:annotation>
                            <xs:documentation>Identifies the name of the data store to use for persistent timers.</xs:documentation>
                        </xs:annotation>
                    </xs:element>
                    <xs:sequence>
                        <xs:element name="persistent-timer-management" type="xs:string">
                            <xs:annotation>
                                <xs:documentation>Identifies the timer-management provider to use for persistent timers.</xs:documentation>
                            </xs:annotation>
                        </xs:element>
                        <xs:element name="transient-timer-management" type="xs:string">
                            <xs:annotation>
                                <xs:documentation>Identifies the timer-management provider to use for persistent timers.</xs:documentation>
                            </xs:annotation>
                        </xs:element>
                    </xs:sequence>
                </xs:choice>
            </xs:extension>
        </xs:complexContent>
    </xs:complexType>
```
The general idea is that a user would specify either persistence-store-name or persistent-timer-management, but not both.

pferraro replied:
@ehsavoie Why are we adding a new schema version?  Version 2.0 of this schema is already new since WF27.

pferraro replied:
I'll address this separately via https://issues.redhat.com/browse/WFLY-17049

chengfang commented on testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/servlet/jboss-ejb3.xml:
```diff
@@ -1,13 +1,13 @@
+<jboss:ejb-jar
+        xmlns="https://jakarta.ee/xml/ns/jakartaee"
+        xmlns:jboss="urn:jboss:jakartaee:1.0"
         xmlns:s="urn:security:1.1"
```
should this be updated to `urn:security:2.0` to match the new jboss-ejb-security_2_0.xsd?

chengfang commented on testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/security/callerprincipal/jboss-ejb3.xml:
```diff
@@ -21,16 +21,16 @@
+<jboss:ejb-jar
+        xmlns="https://jakarta.ee/xml/ns/jakartaee"
+        xmlns:jboss="urn:jboss:jakartaee:1.0"
         xmlns:s="urn:security:1.1"
```
should this be updated to urn:security:2.0 to match the new jboss-ejb-security_2_0.xsd?

chengfang commented on testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/iiop/naming/jboss-ejb3.xml:
```diff
@@ -1,11 +1,11 @@
 <?xml version="1.0" encoding="UTF-8"?>
 
 <jboss:ejb-jar xmlns:jboss="http://www.jboss.com/xml/ns/javaee"
```
change to `urn:jboss:jakartaee:1.0`?

chengfang commented on testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/iiop/naming/jboss-ejb3.xml:
```diff
@@ -1,11 +1,11 @@
+               xsi:schemaLocation="http://www.jboss.com/xml/ns/javaee https://www.jboss.org/schema/jbossas/jboss-ejb3-4_0.xsd
+                  https://jakarta.ee/xml/ns/jakartaee https://www.jboss.org/schema/jbossas/jboss-ejb3-spec-4_0.xsd
                   urn:iiop jboss-ejb-iiop_1_0.xsd"
                version="3.1"
```
change to 4.0?

chengfang commented on testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/iiop/naming/jboss-ejb3.xml:
```diff
@@ -1,11 +1,11 @@
+                  https://jakarta.ee/xml/ns/jakartaee https://www.jboss.org/schema/jbossas/jboss-ejb3-spec-4_0.xsd
                   urn:iiop jboss-ejb-iiop_1_0.xsd"
                version="3.1"
                impl-version="2.0">
```
I think we can just remove this attribute, since it's no longer supported.

chengfang commented on testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/deployment/classloading/ear/EarManifestDependencyPropagatedTestCase.java:
```diff
@@ -88,10 +88,10 @@ private static Class<?> loadClass(String name) throws ClassNotFoundException {
                 "         xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" \n" +
-                "         xsi:schemaLocation=\"http://java.sun.com/xml/ns/javaee http://java.sun.com/xml/ns/javaee/ejb-jar_3_0.xsd\"\n" +
-                "         version=\"3.0\">\n" +
+                "         xsi:schemaLocation=\"https://jakarta.ee/xml/ns/jakartaee https://jakarta.ee/xml/ns/jakartaee//ejb-jar_4_0.xsd\"\n" +
```
`//ejb-jar_` should be `/ejb-jar_`

chengfang commented on testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/mdb/resourceadapter/jboss-ejb3-relative-path.xml:
```diff
@@ -20,9 +20,9 @@
-<jboss xmlns="http://www.jboss.com/xml/ns/javaee" 
-    xmlns:jee="http://java.sun.com/xml/ns/javaee"
-    xmlns:mdb="urn:resource-adapter-binding">
+<jboss xmlns="urn:jboss:jakartaee:1.0" 
```
root element should be `ejb-jar`

chengfang commented on testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/mdb/resourceadapter/jboss-ejb3.xml:
```diff
@@ -1,8 +1,8 @@
-    xmlns:jee="http://java.sun.com/xml/ns/javaee"
-    xmlns:mdb="urn:resource-adapter-binding"
-    xmlns:security="urn:security">
+<jboss xmlns="urn:jboss:jakartaee:1.0" 
```
root element should be `ejb-jar`

bstansberry commented:
@chengfang Can you withdraw your approval of this? AIUI it's out of date. Removing it will help avoid the PR being merged before your more recent inputs are resolved. Thanks!

chengfang commented:
@bstansberry sure, I'll review the updates once the comments are addressed.

chengfang commented on ejb3/src/main/resources/schema/jboss-ejb-timer-service_3_0.xsd:
```diff
@@ -0,0 +1,60 @@
+                    <xs:annotation>
+                        <xs:documentation>Identifies the timer-management provider to use for persistent timers.</xs:documentation>
+                    </xs:annotation>
+                </xs:element>(
```
trailing (

bstansberry commented:
@ehsavoie This has a checkstyle problem

bstansberry commented:
@ehsavoie @chengfang @pferraro Assuming the ongoing CI on #16117 looks ok I want to merge this tonight for the Beta tag tomorrow. But for final Paul's comment at https://github.com/wildfly/wildfly/pull/16084#discussion_r981642256 sounds right -- the 3.0 xsd should be removed and the new in WF 27 2.0 xsd that Paul just updated should be further updated to the jakarta namespace.
